hazelcast/src/main/java/com/hazelcast/jet/JobStatusListener.java

```diff
@@ -0,0 +1,83 @@
+ *      JobStatus#SUSPENDED_EXPORTING_SNAPSHOT SUSPENDED_EXPORTING_SNAPSHOT}
+ *      upon {@linkplain Job#exportSnapshot exporting snapshot}, all of which
+ *      are reported as user-initiated. Other transitions, such as from {@link
+ *      JobStatus#NOT_RUNNING NOT_RUNNING} to {@link JobStatus#STARTING
+ *      STARTING} or from {@link JobStatus#STARTING STARTING} to {@link
+ *      JobStatus#RUNNING RUNNING}, are not reported as user-initiated.
+ * <li> This listener is not notified for {@link JobStatus#COMPLETING
+ *      COMPLETING} status and observes that the status is transitioned from
+ *      {@link JobStatus#RUNNING RUNNING} to {@link JobStatus#FAILED FAILED}
+ *      upon job {@linkplain Job#cancel termination}.
+ * <li> This listener is automatically deregistered after a {@linkplain
+ *      JobStatus#isTerminal terminal event}.
+ * <li> Events are transmitted asynchronously by default, which may result in
+ *      out-of-order events, or the deregistration of this listener before a
+ *      terminal event is received. To disable asynchronous transmission, system
+ *      property {@value EventServiceImpl#EVENT_SYNC_FREQUENCY_PROP} should be
+ *      set to 1. </ol>
+ *
+ * @see Job#addStatusListener(JobStatusListener)
+ * @see Job#removeStatusListener(UUID)
+ * @since 5.3
+ */
+public interface JobStatusListener {
+    /**
+     * Invoked upon job status change. <ol>
+     * <li> This listener observes that the status is transitioned to {@link
+     *      JobStatus#SUSPENDED SUSPENDED}/{@link JobStatus#FAILED FAILED} upon job
+     *      {@linkplain Job#suspend suspension}/{@linkplain Job#cancel termination},
+     *      {@link JobStatus#NOT_RUNNING NOT_RUNNING} upon job {@linkplain
+     *      Job#resume resume}/{@linkplain Job#restart restart}, and {@link
+     *      JobStatus#SUSPENDED_EXPORTING_SNAPSHOT SUSPENDED_EXPORTING_SNAPSHOT}
+     *      upon {@linkplain Job#exportSnapshot exporting snapshot}, all of which
+     *      are reported as user-initiated. Other transitions, such as from {@link
+     *      JobStatus#NOT_RUNNING NOT_RUNNING} to {@link JobStatus#STARTING
+     *      STARTING} or from {@link JobStatus#STARTING STARTING} to {@link
+     *      JobStatus#RUNNING RUNNING}, are not reported as user-initiated.
+     * <li> This listener is not notified for {@link JobStatus#COMPLETING
+     *      COMPLETING} status and observes that the status is transitioned from
+     *      {@link JobStatus#RUNNING RUNNING} to {@link JobStatus#FAILED FAILED}
+     *      upon job {@linkplain Job#cancel termination}.
+     * <li> This listener is automatically deregistered after a {@linkplain
+     *      JobStatus#isTerminal terminal event}.
+     * <li> Events are transmitted asynchronously by default, which may result in
+     *      out-of-order events, or the deregistration of this listener before a
+     *      terminal event is received. To disable asynchronous transmission, system
```

Comment:
@viliam-durina isn't "disable asynchronous transmission" too much global impact to use job status listener? It causes all events (map events etc.) to be sent in synchronous way and waiting for reply instead of just writing message to the connection. This can have significant impact on performance.
our documentation says that order of events is guaranteed (https://docs.hazelcast.com/hazelcast/5.2/events/global-event-configuration).

Reply:
It's our recommendation to the user. The event system isn't reliable in failure scenarios, that's a fact. It can be made more reliable by enabling this property, but it still isn't 100% reliable, it can lose events. We should be honest about that.

The use case is to monitor jobs for failures mostly. If there are network issues, users should assume jobs might have issues, regardless if an event is delivered or not. If there are no network issues, then the events are reliable and should report other types of issues (such as external service down, NPE on bad input etc.). 

We can update the text along these lines.